agent/src/capability/baseconfigstore.cpp

```diff
@@ -23,173 +23,190 @@ namespace softwarecontainer {
-        return ReturnCode::FAILURE;
+        errorMessage = "Configuration is not a json object: "
+            + filePath + ":" + std::to_string(error.line) + " : " + std::string(error.text);
+        log_error() << errorMessage;
+        throw ServiceManifestParseError(errorMessage);
     }
 
     json_t *capabilities = json_object_get(fileroot, "capabilities");
     if (nullptr == capabilities) {
-        log_error() << "Could not parse Capabilities";
-        return ReturnCode::FAILURE;
+        errorMessage = "Could not parse Capabilities in file: " + filePath;
+        log_error() << errorMessage;
+        throw CapabilityParseError(errorMessage);
     }
     if (!json_is_array(capabilities)) {
-        log_error() << "Capabilities is not an array";
-        return ReturnCode::FAILURE;
+        errorMessage = "Capabilities is not an array, in file: " + filePath;
+        log_error() << errorMessage;
+        throw CapabilityParseError(errorMessage);
     }
 
     // Can't use json_decref on fileroot, it removes objects too early
-
-    return parseCapabilities(capabilities);
+    try {
+        parseCapabilities(capabilities);
+    } catch(CapabilityParseError &error) {
+        log_error() << "Could not parse all Capabilities";
+        throw;
+    }
 }
 
-ReturnCode BaseConfigStore::parseCapabilities(json_t *capabilities)
+void BaseConfigStore::parseCapabilities(json_t *capabilities)
 {
     size_t i;
     json_t *capability;
+    std::string errorMessage;
 
     log_debug() << "Size of capabilities is " << std::to_string(json_array_size(capabilities));
 
     json_array_foreach(capabilities, i, capability) {
         if (!json_is_object(capability)) {
-            log_error() << "Capability is not a json object";
-            return ReturnCode::FAILURE;
+            errorMessage = "Capability is not a json object";
+            log_error() << errorMessage;
+            throw CapabilityParseError(errorMessage);
         }
 
         std::string capName;
         if (!JSONParser::read(capability, "name", capName)) {
-            log_error() << "Could not read Capability name";
-            return ReturnCode::FAILURE;
+            errorMessage = "Could not read Capability name";
+            log_error() << errorMessage;
+            throw CapabilityParseError(errorMessage);
         }
 
         json_t *gateways = json_object_get(capability, "gateways");
         if (nullptr == gateways) {
-            log_error() << "Could not read Gateways";
-            return ReturnCode::FAILURE;
+            errorMessage = "Could not read Gateways";
+            log_error() << errorMessage;
+            throw CapabilityParseError(errorMessage);
         }
         if (!json_is_array(gateways)) {
-            log_error() << "Gateways is not an array";
-            return ReturnCode::FAILURE;
+            errorMessage = "Gateways is not an array";
+            log_error() << errorMessage;
+            throw CapabilityParseError(errorMessage);
```

Comment:
You could consider making log_error() + throw SomeError into a function to avoid duplication everywhere. `logAndThrow()` perhaps?